Adjust `toScopeState` to pass the `adsSetupComplete` parameter from callers.

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

It's a follow-up from https://github.com/woocommerce/google-listings-and-ads/pull/1067#discussion_r739221175

- Adjust `toScopeState` to pass the `adsSetupComplete` parameter from callers.
- Add unit tests for `toScopeState` function.

### Detailed test instructions:

1. Run `npm run test-unit -- js/src/utils/toScopeState.test.js`
2. All test cases should pass.

### Changelog entry

eason9487 committed:
Adjust `toScopeState` to pass the `adsSetupComplete` parameter from the consumer side.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1067#discussion_r739221175

eason9487 committed:
Add test cases for  function

ecgan commented on js/src/utils/toScopeState.test.js:
```diff
@@ -0,0 +1,156 @@
+ */
+import toScopeState from './toScopeState';
+
+const SCOPE = {
+	// Manage product listings and accounts for Google Shopping
+	CONTENT: 'https://www.googleapis.com/auth/content',
+	// Manage new site verifications with Google
+	SITE_VERIFICATION_VERIFY_ONLY:
+		'https://www.googleapis.com/auth/siteverification.verify_only',
+	// Manage AdWords campaigns
+	AD_WORDS: 'https://www.googleapis.com/auth/adwords',
+};
```
❔ This is the same as the following, I think we should refactor them? (And also use `Object.freeze` like how I mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/1060#discussion_r736397646 😂 )?

https://github.com/woocommerce/google-listings-and-ads/blob/0cad7cd8defcfbd433f91ea963b12820cba41f24/js/src/utils/toScopeState.js#L6-L14

eason9487 replied:
Sorry, I don't get your point. In current ES practice, even if an object is declared with `const` at the top scope of module/global, devs usually don't mutate its properties without checking the side effects, and almost never saw a dev mutate the properties of an all caps variable name.

If we were developing a library and had to expose constants to third parties for importing, or if we were considering some particular performance tuning, then I think it's a good practice to apply `Object.freeze`. Otherwise, in an app development context, I think it should be ok to just declare a constant and protect it by common conventions. But I may miss something, do you have any concerns?

ecgan replied:
You have a point above. I'm not particularly concerned about `Object.freeze` above, so we can put that aside.

My main concern here is that since now there are two same  `SCOPE` object in `toScopeState.js` and `toScopeState.test.js`, do we want to consider refactor them out?

(The `Object.freeze` suggestion is a secondary concern to the above main concern; if we are not going to refactor them out, then we don't need to use `Object.freeze`.)

ecgan replied:
I looked deeper into the code and figured that the `scopes` parameter in `toScopeState` accepts an array of strings, so the `SCOPE` object in `toScopeState.test.js` is more like a convenience object for our testing here, i.e. we can choose to not have duplicated `SCOPE` object and instead use string array like `['https://www.googleapis.com/auth/content']`, and the tests would still work.

In other words, I think we don't need to refactor this. 😆

eason9487 replied:
The `SCOPE` in *toScopeState.js* file are the expected values. And the `SCOPE` in *test* file are used to simulate the `scopes` array returned from Google API. I think they should not be the same source, otherwise, if the `SCOPE` values in *toScopeState.js* file are changed accidentally, the tests could still pass.

I adjusted it by https://github.com/woocommerce/google-listings-and-ads/pull/1071/commits/1c88e84056e8668ef124ecc24b0f80cd4ee821ba to make them look different and added a comment to avoid confusion.

ecgan replied:
> The `SCOPE` in _toScopeState.js_ file are the expected values. And the `SCOPE` in _test_ file are used to simulate the `scopes` array returned from Google API. I think they should not be the same source, otherwise, if the `SCOPE` values in _toScopeState.js_ file are changed accidentally, the tests could still pass.

Yeah, agree 👍  My mind was playing tricks on me when I saw the same `SCOPE` objects. Sorry about that. 😅

ecgan commented on js/src/utils/toScopeState.test.js:
```diff
@@ -0,0 +1,156 @@
+};
+
+describe( 'toScopeState', () => {
+	let contentScopes;
+	let siteVerificationScopes;
+	let gmcScopes;
+	let adsScopes;
+	let allScopes;
+
+	beforeEach( () => {
+		contentScopes = [ SCOPE.CONTENT ];
+		siteVerificationScopes = [ SCOPE.SITE_VERIFICATION_VERIFY_ONLY ];
+		gmcScopes = [ SCOPE.CONTENT, SCOPE.SITE_VERIFICATION_VERIFY_ONLY ];
+		adsScopes = [ SCOPE.AD_WORDS ];
+		allScopes = [
+			SCOPE.CONTENT,
+			SCOPE.SITE_VERIFICATION_VERIFY_ONLY,
+			SCOPE.AD_WORDS,
+		];
+	} );
```
❔ How about this without needing `beforeEach`? 🤔 

```suggestion
	const contentScopes = [ SCOPE.CONTENT ];
	const siteVerificationScopes = [ SCOPE.SITE_VERIFICATION_VERIFY_ONLY ];
	const gmcScopes = [ SCOPE.CONTENT, SCOPE.SITE_VERIFICATION_VERIFY_ONLY ];
	const adsScopes = [ SCOPE.AD_WORDS ];
	const allScopes = [
		SCOPE.CONTENT,
		SCOPE.SITE_VERIFICATION_VERIFY_ONLY,
		SCOPE.AD_WORDS,
	];
```

ecgan replied:
Note: if we do this, we can even move them to the top level, outside of the `describe` block.

eason9487 replied:
Quoted from https://github.com/woocommerce/google-listings-and-ads/pull/834#discussion_r663177789:

> Maybe I'm paranoid, but I'd move those into `beforeEach` as we would re-use it in every test

This style is based on Tomek's suggestion, but I don't have a particular preference for it. Perhaps you could have a 📜  with @tomalec, and sum up a team convention. 🤔

tomalec replied:
I don't have really strong arguments for it. 
Personally, I like to reason about it like:
- If I need something to happen/to be used for `it` blocks within `describe`. Like initialize variables, define helper function, define "cases", to programmatically generate test, etc.
	I put that inside `describe`. For me, that's something needed to construct `it` blocks, not the actual tests, test's contents.
- If I need something to be prepared for actual tests but once for all. Like, to start a mock server.
	I use `before`. However, I use it relatively rarely.
- If I set up or need something to be prepared for every test content, for every tested behavior. Like, initialize properties/state for a component, mock individual response. So something that will be used **inside** the `it` block.
	I use `beforeEach`.

I believe that makes it easier to reason about the code, as you can anticipate, how the variable & co. will be used just by their location, without inspecting every usage below.
It makes the tests easier to maintain as if you add or remove `it` block, you would not have to adjust the scope of the setup code.
It makes tests more stable, as it helps to make the test less inter-connected and inter-dependent. So one test is not able to mess with the setup of another one.

FWIW, I'm happy to continue that discussion into a separate issue :scroll:.

ecgan commented on js/src/utils/toScopeState.test.js:
```diff
@@ -0,0 +1,156 @@
+				).toBe( true );
+			} );
+
+			it( "and the `scopes` doesn't contains GMC required scopes, should be `false`", () => {
```
💅  ❔ I'd prefer to make the string connects with the `it` in front, so that it reads like "it returns false when...": 😄 

```suggestion
			it( "returns false when the `scopes` doesn't contains GMC required scopes", () => {
```

eason9487 replied:
Quoted from https://github.com/woocommerce/google-listings-and-ads/pull/834#issuecomment-874350184:

> 2\. 💅  Personally, I prefer the test descriptions where the conditions come first, and the expectation is last. I find it easier to write especially in nested `describe`s and more importantly easier to read when the full desc is build:
> 
> ```js
> describe( `When the path…`, () => {
> 	describe( 'for successful…', () => {
> 		it( 'should call …', () => {
> ```

This description structure is based on Tomek's suggestion as well, but I don't have a particular preference for it either. Maybe another 📜  for you and @tomalec. 😂

tomalec replied:
That seems like two perspectives:
- Make the string connects with the `it` in front, so that it [code] reads like "it returns false when..."
- Make the result description build a story, so that test run/failure reads like "When the conditions are met, the subject should return false"

So would we target the code to be more readable, or the test result to be readable. Preferably, we would have both nice and readable :) I'm open to suggestions how to phrase that to make it so.

If we are not able to optimize for both, personally, I'd favor the latter, as the tests are meant to be written once and run every time. Then if they break, it would be easier to reason what and why was broken. Because when it breaks than most probably the person looking at the broken test result is the one who broke it, so the one who is not so familiar with how it should behave.

FWIW, I'm happy to discuss it in a separate issue, and maybe eventually add some conventions and guidelines we agreed on in `.md` file or wiki.

ecgan commented on js/src/utils/toScopeState.test.js:
```diff
@@ -0,0 +1,156 @@
+	} );
+
+	describe.each( [ [ false ], [ true ] ] )(
+		'For `gmcRequired`, if the parameter `adsSetupComplete` = %p',
```
💅  ❔ I'd prefer to make the string connects with the `describe` in front, so that it reads like "describe the returned `gmcRequired` value...". 😄 

```suggestion
		'returned `gmcRequired` value when parameter `adsSetupComplete` = %p'
```

eason9487 replied:
Same as https://github.com/woocommerce/google-listings-and-ads/pull/1071#discussion_r740716477.

ecgan commented on js/src/utils/toScopeState.test.js:
```diff
@@ -0,0 +1,156 @@
+				expect(
+					toScopeState( adsSetupComplete, contentScopes ).gmcRequired
+				).toBe( false );
+				expect(
+					toScopeState( adsSetupComplete, siteVerificationScopes )
+						.gmcRequired
+				).toBe( false );
+				expect(
+					toScopeState( adsSetupComplete, adsScopes ).gmcRequired
+				).toBe( false );
+			} );
+		}
+	);
+
+	describe.each( [ [ false ], [ true ] ] )(
+		'For `adsRequired`, if the parameter `adsSetupComplete` = %p',
+		( adsSetupComplete ) => {
+			it( 'and the `scopes` contains Google Ads required scopes, should be `true`', () => {
+				expect(
+					toScopeState( adsSetupComplete, adsScopes ).adsRequired
+				).toBe( true );
+			} );
+
+			it( "and the `scopes` doesn't contains Google Ads required scopes, should be `false`", () => {
+				expect( toScopeState( adsSetupComplete, [] ).adsRequired ).toBe(
+					false
+				);
+				expect(
+					toScopeState( adsSetupComplete, contentScopes ).adsRequired
+				).toBe( false );
+				expect(
+					toScopeState( adsSetupComplete, siteVerificationScopes )
+						.adsRequired
+				).toBe( false );
+				expect(
+					toScopeState( adsSetupComplete, gmcScopes ).adsRequired
+				).toBe( false );
+			} );
+		}
+	);
```
❔ By looking at these two `describe.each` blocks here, we would notice that for `gmcRequired` and `adsRequired`, the `toScopeState`'s first parameter `adsSetupComplete` - whether it is `true` or `false` - does not matter. 

I feel like the API isn't really that nice, and there should be a better way. 🤔 

I looked into the application code. We don't have any usage of `adsRequired`. We have one usage of `gmcRequired` which is used in the following...

https://github.com/woocommerce/google-listings-and-ads/blob/0cad7cd8defcfbd433f91ea963b12820cba41f24/js/src/setup-mc/setup-stepper/setup-accounts/index.js#L40-L42

... which I think can be replaced with `glaRequired` and it would work fine, which means that `gmcRequired` is not needed too. 🤔 

If `gmcRequired` and `adsRequired` are not needed, then we can simplify `toScopeState` (and its tests), e.g. instead of `toScopeState (arg1, arg2)` which returns an object, it can be `hasRequiredPermissions (arg1, arg2)` which returns a boolean value (i.e. `glaRequired`).

What do you think?

eason9487 replied:
`adsRequired` is used [here](https://github.com/woocommerce/google-listings-and-ads/commit/523fb30191e63c64d772c39e8a0efc49e607854a#diff-71606b9acaa50832ca8d6c49bb920d55cd2a1c2e02f13ac9028f4f468b0c5d9eR19) (PR #1069)

ecgan replied:
> `adsRequired` is used [here](https://github.com/woocommerce/google-listings-and-ads/commit/523fb30191e63c64d772c39e8a0efc49e607854a#diff-71606b9acaa50832ca8d6c49bb920d55cd2a1c2e02f13ac9028f4f468b0c5d9eR19) (PR #1069)

That `adsRequired` can be replaced with `glaRequired` and things would still work the same too, right? 🤔

eason9487 replied:
Replace `adsRequired` with `glaRequired` could get the correct results.

However, the semantics are not consistent with the purpose, because on the Google Ads setup page, the only permission that needs to be checked is the ad permission.

And even if the replacement works correctly at this moment, it'd hide a potential problem that if new features are added that require new permissions and also added into `glaRequired`, then the logic [here](https://github.com/woocommerce/google-listings-and-ads/commit/523fb30191e63c64d772c39e8a0efc49e607854a#diff-71606b9acaa50832ca8d6c49bb920d55cd2a1c2e02f13ac9028f4f468b0c5d9eR19) might become incorrect. I lean toward keeping it as is.

ecgan reviewed (approved):
Tested and working as expected. Left 5 comments below, which I think could improve and simplify the code. Approving this since there is nothing major blocking us. 👍

eason9487 committed:
Create the simulated input parameter `scopes` in another way

Address https://github.com/woocommerce/google-listings-and-ads/pull/1071#discussion_r740199990
